examples/Prioritized_DQN/per/utils.py

```diff
@@ -0,0 +1,208 @@
+            if rightpos < endpos and \
+                not self.idx_heap[childpos] > self.idx_heap[rightpos]:
+                childpos = rightpos
+            self.idx_heap[pos] = self.idx_heap[childpos]
+            self._idx_pos[self.idx_heap[childpos][1]] = pos
+            pos = childpos
+            childpos = 2 * pos + 1
+        self.idx_heap[pos] = item
+        self._idx_pos[item[1]] = pos
+        self._siftdown(startpos, pos)
+
+    def from_list(self, lst):
+        assert len(lst) == self.max_size
+        self.elements = lst.copy()
+        for idx, elem in enumerate(self.elements):
+            self.idx_heap.append([1.0, idx])
+            self._idx_pos[idx] = idx
+
+
+class SumTree:
+    def __init__(self, capacity):
+        self.capacity = capacity
+        self.elements = [None for _ in range(capacity)]
+        self.tree = [0 for _ in range(2 * capacity - 1)]
+        self._ptr = 0
+
+    def full(self):
+        return all(self.elements)  # no `None` in self.elements
+
+    def add(self, item, priority):
+        self.elements[self._ptr] = item
+        tree_idx = self._ptr + self.capacity - 1
+        self.update(tree_idx, priority)
+        self._ptr = (self._ptr + 1) % self.capacity
+
+    def update(self, tree_idx, priority):
+        diff = priority - self.tree[tree_idx]
+        self.tree[tree_idx] = priority
+        while tree_idx != 0:
+            tree_idx = (tree_idx - 1) >> 1
+            self.tree[tree_idx] += diff
+
+    def retrieve(self, value):
+        parent_idx = 0
+        while True:
+            left_child_idx = 2 * parent_idx + 1
+            right_child_idx = left_child_idx + 1
+            if left_child_idx >= len(self.tree):
+                leaf_idx = parent_idx
+                break
+            else:
+                if value <= self.tree[left_child_idx]:
+                    parent_idx = left_child_idx
+                else:
+                    value -= self.tree[left_child_idx]
+                    parent_idx = right_child_idx
+        elem_idx = leaf_idx - self.capacity + 1
+        priority = self.tree[leaf_idx]
+        return self.elements[elem_idx], leaf_idx, priority
+
+    def from_list(self, lst):
+        assert len(lst) == self.capacity
+        self.elements = lst.copy()
+        for i in range(self.capacity - 1, 2 * self.capacity - 1):
+            self.update(i, 1.0)
+
+    @property
+    def total_p(self):
+        return self.tree[0]
+
+
+class BasePER:
+    def __init__(self,
+                 alpha,
+                 seg_num,
+                 elem_cls,
+                 size=1e6,
+                 init_mem=None,
+                 framestack=4):
+        self.alpha = alpha
+        self.seg_num = seg_num
+        self.size = int(size)
+        self.elements = elem_cls(self.size)
+        if init_mem:
+            self.elements.from_list(init_mem)
+        self.framestack = framestack
+        self._max_priority = 1.0
+
+    def _check_full(self):
```

Comment:
The implementation is not compact, maybe can replace it with 
`assert self.elements.full(), "The replay memory is not full!"`